File: phases/04-algorithms-and-complexity/28-parallel-algorithms-pram-brent-map-reduce-style/code/main.py
"""
Parallel Algorithms — PRAM, Brent, Map-Reduce style
Phase 04 — Algorithms & Complexity Analysis

Parallel prefix sum, parallel merge sort, speedup benchmarks.
"""

import math
import time
from concurrent.futures import ThreadPoolExecutor


# ─── Parallel Prefix Sum (Blelloch-style scan) ────────────────────────────


def sequential_prefix_sum(arr: list[int]) -> list[int]:
    """Exclusive prefix sum: out[i] = sum of arr[0..i)."""
    out = []
    acc = 0
    for x in arr:
        out.append(acc)
        acc += x
    return out


def parallel_prefix_sum(arr: list[int]) -> list[int]:
    """Work-efficient exclusive prefix sum using up-sweep + down-sweep.

    Work:  O(n)       (matches sequential)
    Depth: O(log n)   (log n tree levels)

    Uses ThreadPoolExecutor to simulate PRAM concurrent access.
    """
    n = len(arr)
    if n == 0:
        return []
    if n == 1:
        return [0]

    # Pad to next power of two
    m = 1 << math.ceil(math.log2(n))
    buf = list(arr) + [0] * (m - n)

    # ── Up-sweep (reduce) ──
    d = 1
    while d < m:
        pairs = [(i, d) for i in range(d, m, 2 * d)]

        def _up(pair, _buf=buf):
            i, step = pair
            _buf[i + step - 1] += _buf[i - 1]

        with ThreadPoolExecutor() as pool:
            list(pool.map(_up, pairs))
        d *= 2

    # Set last element to 0 (exclusive scan)
    buf[m - 1] = 0

    # ── Down-sweep ──
    d = m // 2
    while d >= 1:
        pairs = [(i, d) for i in range(d, m, 2 * d)]

        def _down(pair, _buf=buf):
            i, step = pair
            old_left = _buf[i - 1]
            _buf[i - 1] = _buf[i + step - 1]
            _buf[i + step - 1] += old_left

        with ThreadPoolExecutor() as pool:
            list(pool.map(_down, pairs))
        d //= 2

    return buf[:n]


# ─── Parallel Merge Sort ──────────────────────────────────────────────────


def sequential_merge(left: list[int], right: list[int]) -> list[int]:
    result = []
    i = j = 0
    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1
    result.extend(left[i:])
    result.extend(right[j:])
    return result


def parallel_merge_sort(
    arr: list[int],
    executor: ThreadPoolExecutor | None = None,
    depth: int = 0,
    max_parallel_depth: int = 4,
) -> list[int]:
    """Fork-join parallel merge sort.

    Work:  O(n log n)  (same as sequential merge sort)
    Depth: O(n)        (naive sequential merge); O(log^2 n) with parallel merge.

    Limits fork depth to max_parallel_depth to avoid thread explosion.
    """
    if len(arr) <= 1:
        return arr

    own_executor = executor is None
    if own_executor:
        executor = ThreadPoolExecutor()

    mid = len(arr) // 2

    try:
        if depth < max_parallel_depth:
            left_fut = executor.submit(
                parallel_merge_sort, arr[:mid], executor, depth + 1, max_parallel_depth
            )
            right_fut = executor.submit(
                parallel_merge_sort, arr[mid:], executor, depth + 1, max_parallel_depth
            )
            left = left_fut.result()
            right = right_fut.result()
        else:
            left = sorted(arr[:mid])
            right = sorted(arr[mid:])
    finally:
        if own_executor:
            executor.shutdown(wait=True)

    return sequential_merge(left, right)


# ─── Work-Depth Analysis ──────────────────────────────────────────────────


def prefix_sum_work_depth(n: int) -> dict:
    """Return theoretical work and depth for parallel prefix sum."""
    return {
        "work": f"O({n})",
        "depth": f"O({int(math.log2(max(n, 1)))})",
        "work_expr": f"O(n) = O({n})",
        "depth_expr": f"O(log n) = O({int(math.log2(max(n, 1)))})",
    }


def merge_sort_work_depth(n: int) -> dict:
    """Return theoretical work and depth for parallel merge sort (naive merge)."""
    log_n = int(math.log2(max(n, 1)))
    return {
        "work": f"O({n} log {n})",
        "depth": f"O({n})  [sequential merge]; O({log_n}^2) with parallel merge",
        "work_expr": f"O(n log n) ~ O({n * log_n if n > 1 else 0})",
        "depth_expr": f"O(log^2 n) = O({log_n ** 2}) [parallel merge]",
    }


# ─── Speedup Measurement ──────────────────────────────────────────────────


def measure_speedup(
    parallel_fn,
    sequential_fn,
    data: list[int],
    label: str = "",
) -> tuple[float, float]:
    t0 = time.perf_counter()
    seq_result = sequential_fn(data)
    t_seq = time.perf_counter() - t0

    t0 = time.perf_counter()
    par_result = parallel_fn(data)
    t_par = time.perf_counter() - t0

    assert par_result == seq_result, f"Results differ!\n  seq={seq_result}\n  par={par_result}"

    speedup = t_seq / t_par if t_par > 0 else float("inf")
    print(f"  [{label}]")
    print(f"    Sequential: {t_seq:.6f}s")
    print(f"    Parallel:   {t_par:.6f}s")
    print(f"    Speedup:    {speedup:.2f}x")
    print()

    return t_seq, t_par


# ─── Map-Reduce Style Demo ────────────────────────────────────────────────


def map_reduce_word_count(lines: list[str]) -> dict[str, int]:
    """Classic Map-Reduce word count using Python's thread pool.

    Map:    each line → list of (word, 1) pairs  [parallel]
    Shuffle: group by word                        [sequential in this demo]
    Reduce: sum counts per word                   [parallel]
    """

    # Map phase
    def mapper(line: str) -> list[tuple[str, int]]:
        return [(w.lower(), 1) for w in line.split() if w]

    with ThreadPoolExecutor() as pool:
        mapped = list(pool.map(mapper, lines))

    # Shuffle phase
    groups: dict[str, list[int]] = {}
    for pairs in mapped:
        for word, count in pairs:
            groups.setdefault(word, []).append(count)

    # Reduce phase
    def reducer(item: tuple[str, list[int]]) -> tuple[str, int]:
        word, counts = item
        return word, sum(counts)

    with ThreadPoolExecutor() as pool:
        results = list(pool.map(reducer, groups.items()))

    return dict(results)


# ─── main ─────────────────────────────────────────────────────────────────


def main() -> None:
    import random

    random.seed(42)

    # ── Prefix Sum ──
    print("=== Parallel Prefix Sum ===")
    for size_exp in [8, 12, 16]:
        n = 2 ** size_exp
        data = [random.randint(1, 100) for _ in range(n)]
        ad = prefix_sum_work_depth(n)
        print(f"  n={n}: work={ad['work']}, depth={ad['depth']}")
        measure_speedup(parallel_prefix_sum, sequential_prefix_sum, data, label=f"prefix_sum n={n}")

    # ── Merge Sort ──
    print("=== Parallel Merge Sort ===")
    for size_exp in [10, 14]:
        n = 2 ** size_exp
        data = [random.randint(1, 10_000) for _ in range(n)]
        ad = merge_sort_work_depth(n)
        print(f"  n={n}: work={ad['work']}, depth={ad['depth']}")
        measure_speedup(
            lambda arr: parallel_merge_sort(arr),
            sorted,
            data,
            label=f"merge_sort n={n}",
        )

    # ── Map-Reduce Word Count ──
    print("=== Map-Reduce Word Count ===")
    lines = [
        "the quick brown fox jumps over the lazy dog",
        "the lazy dog sleeps under the brown tree",
        "a quick fox and a lazy dog play together",
    ]
    counts = map_reduce_word_count(lines)
    for word in sorted(counts):
        print(f"  {word}: {counts[word]}")

    # ── Brent's Theorem Demo ──
    print("\n=== Brent's Theorem ===")
    T1 = 1000   # work (sequential time)
    T_inf = 10  # depth (critical path)
    for p in [1, 2, 4, 8, 16, 32, 64]:
        lower = math.ceil(T1 / p)
        upper = math.ceil(T1 / p) + T_inf
        print(f"  p={p:3d}:  T1/p = {lower:4d}  ≤  T_p  ≤  {upper:4d}  (= T1/p + T∞)")


if __name__ == "__main__":
    main()

File: phases/04-algorithms-and-complexity/28-parallel-algorithms-pram-brent-map-reduce-style/code/main.rs
//! Parallel Algorithms — PRAM, Brent, Map-Reduce style
//! Phase 04 — Algorithms & Complexity Analysis
//!
//! Parallel prefix sum, parallel merge sort, speedup benchmarks
//! using std::thread (no external crates).

use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use std::thread;
use std::time::Instant;

// ─── Parallel Prefix Sum (Blelloch scan) ──────────────────────────────────

/// Exclusive prefix sum using up-sweep + down-sweep.
///
/// Work:  O(n)
/// Depth: O(log n)
///
/// Uses std::thread to simulate PRAM concurrent access.
fn parallel_prefix_sum(arr: &[i64]) -> Vec<i64> {
    let n = arr.len();
    if n == 0 {
        return vec![];
    }
    if n == 1 {
        return vec![0];
    }

    let m = n.next_power_of_two();
    let buf: Arc<Mutex<Vec<i64>>> = Arc::new(Mutex::new(
        arr.iter().copied().chain(std::iter::repeat(0)).take(m).collect(),
    ));

    // ── Up-sweep (reduce) ──
    let mut d: usize = 1;
    while d < m {
        let step = d;
        let indices: Vec<usize> = (d..m).step_by(2 * d).collect();
        let mut handles = Vec::new();
        for &i in &indices {
            let buf = Arc::clone(&buf);
            handles.push(thread::spawn(move || {
                let mut b = buf.lock().unwrap();
                b[i + step - 1] += b[i - 1];
            }));
        }
        for h in handles {
            h.join().unwrap();
        }
        d *= 2;
    }

    // Exclusive scan: set root to 0
    {
        let mut b = buf.lock().unwrap();
        b[m - 1] = 0;
    }

    // ── Down-sweep ──
    let mut d = m / 2;
    while d >= 1 {
        let step = d;
        let indices: Vec<usize> = (d..m).step_by(2 * d).collect();
        let mut handles = Vec::new();
        for &i in &indices {
            let buf = Arc::clone(&buf);
            handles.push(thread::spawn(move || {
                let mut b = buf.lock().unwrap();
                let old_left = b[i - 1];
                b[i - 1] = b[i + step - 1];
                b[i + step - 1] += old_left;
            }));
        }
        for h in handles {
            h.join().unwrap();
        }
        d /= 2;
    }

    let b = buf.lock().unwrap();
    b[..n].to_vec()
}

fn sequential_prefix_sum(arr: &[i64]) -> Vec<i64> {
    let mut out = Vec::with_capacity(arr.len());
    let mut acc: i64 = 0;
    for &x in arr {
        out.push(acc);
        acc += x;
    }
    out
}

// ─── Parallel Merge Sort ──────────────────────────────────────────────────

fn sequential_merge(left: &[i64], right: &[i64]) -> Vec<i64> {
    let mut result = Vec::with_capacity(left.len() + right.len());
    let mut i = 0;
    let mut j = 0;
    while i < left.len() && j < right.len() {
        if left[i] <= right[j] {
            result.push(left[i]);
            i += 1;
        } else {
            result.push(right[j]);
            j += 1;
        }
    }
    result.extend_from_slice(&left[i..]);
    result.extend_from_slice(&right[j..]);
    result
}

/// Fork-join parallel merge sort using std::thread.
///
/// Work:  O(n log n)
/// Depth: O(log^2 n) with parallel merge (sequential merge here → O(n) depth).
///
/// Limits recursion depth to avoid thread explosion.
fn parallel_merge_sort(arr: &[i64], depth: usize, max_parallel_depth: usize) -> Vec<i64> {
    if arr.len() <= 1 {
        return arr.to_vec();
    }

    let mid = arr.len() / 2;

    if depth < max_parallel_depth {
        let left_arr = arr[..mid].to_vec();
        let right_arr = arr[mid..].to_vec();

        let left_handle = thread::spawn(move || parallel_merge_sort(&left_arr, depth + 1, max_parallel_depth));
        let right_handle = thread::spawn(move || parallel_merge_sort(&right_arr, depth + 1, max_parallel_depth));

        let left = left_handle.join().unwrap();
        let right = right_handle.join().unwrap();

        sequential_merge(&left, &right)
    } else {
        let mut sorted = arr.to_vec();
        sorted.sort();
        sorted
    }
}

// ─── Speedup Measurement ─────────────────────────────────────────────────

fn measure_speedup<F, G, R>(label: &str, parallel_fn: F, sequential_fn: G, data: &[i64])
where
    F: Fn(&[i64]) -> R,
    G: Fn(&[i64]) -> R,
    R: PartialEq + std::fmt::Debug,
{
    let t0 = Instant::now();
    let seq_result = sequential_fn(data);
    let t_seq = t0.elapsed();

    let t0 = Instant::now();
    let par_result = parallel_fn(data);
    let t_par = t0.elapsed();

    assert_eq!(seq_result, par_result, "Results differ!");

    let speedup = if t_par.as_secs_f64() > 0.0 {
        t_seq.as_secs_f64() / t_par.as_secs_f64()
    } else {
        f64::INFINITY
    };

    println!("  [{}]", label);
    println!("    Sequential: {:.6}s", t_seq.as_secs_f64());
    println!("    Parallel:   {:.6}s", t_par.as_secs_f64());
    println!("    Speedup:    {:.2}x", speedup);
    println!();
}

// ─── Map-Reduce Style ────────────────────────────────────────────────────

/// Map-Reduce word count using threads.
///
/// Map:    each line → (word, 1) pairs  [parallel]
/// Shuffle: group by word
/// Reduce: sum counts per word          [parallel via Arc<Mutex>
fn map_reduce_word_count(lines: &[&str]) -> HashMap<String, i64> {
    // Map phase
    let mut handles = Vec::new();
    for &line in lines {
        handles.push(thread::spawn(move || {
            let mut pairs: Vec<(String, i64)> = Vec::new();
            for word in line.split_whitespace() {
                pairs.push((word.to_lowercase(), 1));
            }
            pairs
        }));
    }

    let mut all_pairs: Vec<(String, i64)> = Vec::new();
    for h in handles {
        all_pairs.extend(h.join().unwrap());
    }

    // Shuffle phase
    let mut groups: HashMap<String, Vec<i64>> = HashMap::new();
    for (word, count) in all_pairs {
        groups.entry(word).or_default().push(count);
    }

    // Reduce phase
    let groups_arc = Arc::new(Mutex::new(groups));
    let result = Arc::new(Mutex::new(HashMap::new()));
    let keys: Vec<String> = {
        let g = groups_arc.lock().unwrap();
        g.keys().cloned().collect()
    };

    let mut handles = Vec::new();
    for key in keys {
        let groups_ref = Arc::clone(&groups_arc);
        let result_ref = Arc::clone(&result);
        handles.push(thread::spawn(move || {
            let g = groups_ref.lock().unwrap();
            let total: i64 = g.get(&key).unwrap().iter().sum();
            let mut r = result_ref.lock().unwrap();
            r.insert(key, total);
        }));
    }
    for h in handles {
        h.join().unwrap();
    }

    Arc::try_unwrap(result).unwrap().into_inner().unwrap()
}

// ─── Brent's Theorem Demo ────────────────────────────────────────────────

fn brent_demo() {
    println!("=== Brent's Theorem ===");
    let t1: f64 = 1000.0;
    let t_inf: f64 = 10.0;

    for &p in &[1, 2, 4, 8, 16, 32, 64] {
        let lower = (t1 / p as f64).ceil() as i64;
        let upper = (t1 / p as f64).ceil() as i64 + t_inf as i64;
        println!(
            "  p={:3}:  T1/p = {:4}  ≤  T_p  ≤  {:4}  (= T1/p + T∞)",
            p, lower, upper
        );
    }
}

// ─── Simple PRNG (no external crate) ─────────────────────────────────────

fn simple_rand(seed: &mut u64) -> i64 {
    *seed = seed.wrapping_mul(6364136223846793005).wrapping_add(1);
    ((*seed >> 33) % 100) as i64 + 1
}

// ─── main ─────────────────────────────────────────────────────────────────

fn main() {
    let mut seed: u64 = 42;

    // ── Prefix Sum ──
    println!("=== Parallel Prefix Sum ===");
    for &size_exp in &[8, 12, 16] {
        let n = 1 << size_exp;
        let data: Vec<i64> = (0..n).map(|_| simple_rand(&mut seed)).collect();
        let log_n = (n as f64).log2() as usize;
        println!("  n={}: work=O({}), depth=O({})", n, n, log_n);

        measure_speedup(
            &format!("prefix_sum n={}", n),
            |arr: &[i64]| parallel_prefix_sum(arr),
            |arr: &[i64]| sequential_prefix_sum(arr),
            &data,
        );
    }

    // ── Merge Sort ──
    println!("=== Parallel Merge Sort ===");
    for &size_exp in &[10, 14] {
        let n = 1 << size_exp;
        let data: Vec<i64> = (0..n).map(|_| simple_rand(&mut seed) * 100).collect();
        let log_n = (n as f64).log2() as usize;
        println!(
            "  n={}: work=O(n log n) ≈ O({}), depth=O(log² n) ≈ O({})",
            n,
            n * log_n,
            log_n * log_n
        );

        measure_speedup(
            &format!("merge_sort n={}", n),
            |arr: &[i64]| parallel_merge_sort(arr, 0, 4),
            |arr: &[i64]| {
                let mut v = arr.to_vec();
                v.sort();
                v
            },
            &data,
        );
    }

    // ── Map-Reduce Word Count ──
    println!("=== Map-Reduce Word Count ===");
    let lines = [
        "the quick brown fox jumps over the lazy dog",
        "the lazy dog sleeps under the brown tree",
        "a quick fox and a lazy dog play together",
    ];
    let counts = map_reduce_word_count(&lines);
    let mut words: Vec<_> = counts.keys().cloned().collect();
    words.sort();
    for w in &words {
        println!("  {}: {}", w, counts[w]);
    }

    // ── Brent's Theorem ──
    println!();
    brent_demo();
}
